Announcing Wasm alpha availability blog. (#10530)
istio/istio.io: content/en/blog/2021/wasm-api-alpha/index.md

---
title: Announcing the alpha availability of WebAssembly Plugins
description: Announcement of the alpha availability of Wasm plugins in Istio.

> publishdate: 2021-11-25

Author: is TBD

{{< text yaml >}}
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: your-filter
spec:
  selector:
    matchLabels:
      app: server
  phase: AUTHN
  priority: 10
  pluginConfig:
    someSetting: true
    someOtherSetting: false
    youNameIt:
    - first
    - second
  url: docker.io/your-org/your-filter:1.0.0
{{< /text >}}

There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.

The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:

AUTHN: this is prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.
AUTHZ: just after the Istio authentication filters and before any authorization filters, i.e. before AuthorizationPolicies have been applied.

> STATS:

Reviewer: Looks like this is incomplete. 4 phases mentioned, auth/authn described . Stats not described.

Author: done! https://github.com/istio/istio.io/pull/10530/commits/082c727c28946aca754ede4287d1c1c287a564f9

> `pluginConfig` is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config on `on_configure` in [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or `OnPluginStart` call back in [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).

Reviewer: pl Add a link to c++ sdk as well.

Author: done https://github.com/istio/istio.io/pull/10530/commits/082c727c28946aca754ede4287d1c1c287a564f9

> The `url` field specifies where to pull the Wasm module. You’ll notice that the `url` in this case is a docker URI - this is because apart from loading Wasm modules via HTTP, HTTPS and the local file system (using file://), we are introducing a container image format to package Wasm extensions for Istio.

Reviewer: We are introducing oci as the preferred mechanism for distributing wasm modules.

Author: done

For detail, please refer to [the link here](https://github.com/solo-io/wasm/tree/master/spec), and learn [how to build OCI images](https://github.com/solo-io/wasm/tree/master/spec#how-can-i-build-images) that are consumable by Istio agent.

## Image fetcher in Istio agent

As of 1.9, Istio-agent provides a reliable solution for loading Wasm binaries fetched from remote HTTP sources configured in the `EnvoyFilters` by leveraging the xDS proxy inside istio-agent and Envoy’s Extension Configuration Discovery Service (ECDS). The same mechanism applies for the new Wasm API implementation in Istio 1.12, and you can use HTTP remote resources reliably without the concern that Envoy might get stuck in bad configurations when the remote fetch fails.

> In addition, Istio 1.12 expands this capability to the Wasm OCI images. That means the Istio agent is now able to fetch Wasm images from any OCI registry including Docker Hub, Google Container Registry(GCR), Amazon Elastic Container Registry (Amazon ECR), etc. After fetching images, Istio-agent extracts and caches Wasm binaries from them, and then inserts them into Envoy filter chains.

Reviewer: Extracts and caches “in memory”

Author: I think it is not in-memory, but it actually writes into the local file system https://github.com/istio/istio/blob/master/pkg/wasm/cache.go#L195-L202

{{< image width="75%"
    link="./istio-agent-architecture.svg"
    alt="Remote Wasm module fetch flow"
    caption="Remote Wasm module fetch flow"
    >}}

## Improvements in Envoy Wasm runtime

The Wasm runtime powered by V8 in Envoy has been shipped as of Istio 1.5, and there has been a lot of improvements since then.

> First of all, some of the WASI (WebAssembly System Interface) system calls are supported now. For example, `clock_time_get` system call can be made from Wasm programs, which means that you can use `std::time::SystemTime::now()` in Rust or `time.Now().UnixNano()` in Go for Envoy Wasm extensions just like any other native platform. Another example is that `random_get` is supported by Envoy now, so the "crypto/rand" package is available in Go as a cryptographically secure random number generator. We are currently actively looking into file system support as we have seen the requests for reading and writing local files from Wasm programs running in Envoy.

Reviewer: This is really good, can we add a subheading for these sub sections ?

WASi, Statckrace, debug

Author: done

  4:  0xea15 - proxy_on_request_headers
{{< /text >}}

The above is an example output of the stack trace for Go SDK based Wasm extensions. You might notice that you want file names and lines in the trace lines -- that is another huge future work and is currently an open issue since it has something to do with the DWARF format for WebAssembly and the Exception Handling proposal to WebAssembly specification.

Also you can see strace equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, we can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example log stream of such strace logs:

{{< text plain >}}
[host->vm] proxy_on_context_create(2, 1)
[host<-vm] proxy_on_context_create return: void
[host->vm] proxy_on_request_headers(2, 8, 1)
[vm->host] wasi_snapshot_preview1.random_get(86928, 32)
[vm<-host] wasi_snapshot_preview1.random_get return: 0
[vm->host] env.proxy_log(2, 87776, 18)
{{< /text >}}

This is especially useful to debug Wasm program’s execution at runtime, for example, to verify it is not making any malicious system calls.

> The next one is about metrics. Wasm extensions have been able to define their own custom metrics and expose them in Envoy just like any other metric, but prior to Istio 1.12, all of these custom metrics are prefixed by `envoy_` Prometheus namespace and users were not be able to have their own namespaces. Now, you can choose whatever namespace you want, and your metrics are exposed in Envoy as-is without being prefixed by `envoy_`.

Reviewer: Link to additional configuration needed to expose it from prom. Stats inclusion etc
@bianpengyuan

Author: hope this is what you meant :D https://github.com/istio/istio.io/pull/10530/commits/5b4eb8a97de1dbc405bd05d972d2a53f1188df13

{{< text yaml >}}
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: your-filter
spec:
  selector:
    matchLabels:
      app: server
  phase: AUTHN
  priority: 10
  pluginConfig:
    someSetting: true
    someOtherSetting: false
    youNameIt:
    - first
    - second
  url: docker.io/your-org/your-filter:1.0.0
{{< /text >}}

There are a lot of similarities and a few differences between WasmPlugin and EnvoyFilter, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an EnvoyFilter.

The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:

AUTHN: this is prior to any istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.
AUTHZ: just after the istio authentication filters and before any authorization filters, i.e. before AuthorizationPolicies have been applied.

> STATS: 

Author: @dgn What should I put on this line?

Author: https://github.com/istio/istio.io/pull/10530/commits/082c727c28946aca754ede4287d1c1c287a564f9 done

---
title: Announcing the alpha availability of WebAssembly Plugins
description: Introduction of Wasm Plugin API, and updates on Envoy and Istio's Wasm based plugins.
publishdate: 2021-11-25

> attribution: "Daniel Grimm(Red Hat), Pengyuan Bian(Google), Takeshi Yoneda(Tetrate)"

Reviewer 2: There is a missing space on the attribution line between the author and the `-`.
```
 By Daniel Grimm- Red Hat, Pengyuan Bian- Google, Takeshi Yoneda- Tetrate
```

keywords: [wasm,extensibility,WebAssembly]
---

> [At the Istio 1.9 release](../wasm-progress/), we introduced experimental support for Wasm module distribution and Wasm extensions ecosystem repository for canonical examples and use cases of Wasm extension development. Over the past 9 months, the Istio, Envoy, and Proxy-Wasm communities have continued our joint efforts to make WebAssembly (Wasm) extensibility stable, reliable, and easy to adopt, and we are pleased to announce Alpha support for Wasm extensibility at Istio 1.12 release! Let’s walk through the updates to Wasm support through the Istio 1.12 release.

Reviewer 2: In the Istio 1.9...    extensibility in the Istio 1.12

The first sentence doesn't make sense to me. To many and's? 

Also make the WebAssembly (Wasm) in the first sentence.

Author: @bianpengyuan wdyt? ^^

## New WasmPlugin API

With the new WasmPlugin API in the `extensions.istio.io` namespace, we’re introducing a new high-level API for extending the functionality of the Istio Proxy with custom WebAssembly (Wasm) modules. This effort builds on the excellent work that has gone into the [Proxy-Wasm](https://github.com/proxy-wasm) specification and implementation over the last two years. From now on, you no longer need to use `EnvoyFilter` resources to add custom Wasm modules to your proxies. Instead, you can now use WasmPlugin directly:

{{< text yaml >}}
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: your-filter
spec:
  selector:
    matchLabels:
      app: server
  phase: AUTHN
  priority: 10
  pluginConfig:
    someSetting: true
    someOtherSetting: false
    youNameIt:
    - first
    - second
  url: docker.io/your-org/your-filter:1.0.0
{{< /text >}}

There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.

> The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:

Reviewer 2: You mention next on a new line, but there was no previous on a new line. I'm thinking you wanted `selector` as the first field, so split it from the prior paragraph?

* `AUTHN`: prior to any Istio authentication and authorization filters.

> * `AUTHZ`: after the Istio authentication filters and before any first class authorization filters, i.e. before AuthorizationPolicies have been applied.

Reviewer 2: first-class

`pluginConfig` is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config on `onConfigure` in [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/fd0be8405db25de0264bdb78fae3a82668c03782/proxy_wasm_api.h#L329-L331), `on_configure` in [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or `OnPluginStart` call back in [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).

The `url` field specifies where to pull the Wasm module. You’ll notice that the `url` in this case is a docker URI - this is because apart from loading Wasm modules via HTTP, HTTPS and the local file system (using file://), we are introducing the OCI image format as the preferred mechanism for distributing Wasm modules.

One last thing to note is currently the Wasm Plugin API only applies to inbound HTTP filter chains.
Adding support for network filters and outbound traffics is the future work.

## Wasm image specification

> We believe that containers are the ideal way to store, publish and manage proxy extensions, so we worked with Solo.io to extend their existing Proxy-Wasm container format with a variant that aims to be compatible with all registries and CLI toolchain. Depending on your processes, you can now build your proxy extension containers using your existing container CLI tooling such as Docker CLI or [buildah](https://buildah.io/).

Reviewer 2: and the CLI toolchain.

> For detail, please refer to [the link here](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md), and learn how to build OCI images that are consumable by Istio agent.

Reviewer 2: Maybe make this something like: 

To learn how to build OCI images, please refer to this [link]().

## Image fetcher in Istio agent

As of 1.9, Istio-agent provides a reliable solution for loading Wasm binaries fetched from remote HTTP sources configured in the `EnvoyFilters` by leveraging the xDS proxy inside istio-agent and Envoy’s Extension Configuration Discovery Service (ECDS). The same mechanism applies for the new Wasm API implementation in Istio 1.12, and you can use HTTP remote resources reliably without the concern that Envoy might get stuck in bad configurations when the remote fetch fails.

In addition, Istio 1.12 expands this capability to the Wasm OCI images. That means the Istio agent is now able to fetch Wasm images from any OCI registry including Docker Hub, Google Container Registry(GCR), Amazon Elastic Container Registry (Amazon ECR), etc. After fetching images, Istio-agent extracts and caches Wasm binaries from them, and then inserts them into Envoy filter chains.

{{< image width="75%"
    link="./istio-agent-architecture.svg"
    alt="Remote Wasm module fetch flow"
    caption="Remote Wasm module fetch flow"
    >}}

## Improvements in Envoy Wasm runtime

> The Wasm runtime powered by V8 in Envoy has been shipped as of Istio 1.5, and there has been a lot of improvements since then.

Reviewer 2: remove the comma.

### WASI supports

> First of all, some of the WASI (WebAssembly System Interface) system calls are supported now. For example, `clock_time_get` system call can be made from Wasm programs, which means that you can use `std::time::SystemTime::now()` in Rust or `time.Now().UnixNano()` in Go for Envoy Wasm extensions just like any other native platform. Another example is that `random_get` is supported by Envoy now, so the "crypto/rand" package is available in Go as a cryptographically secure random number generator. We are currently actively looking into file system support as we have seen the requests for reading and writing local files from Wasm programs running in Envoy.

Reviewer 2: First, ~~of all~~

Reviewer 2: supported now -> now supported.

### Debuggability

> Next is the improvement in debuggability. Now the Envoy runtime emits the stack trace of your program when it causes runtime errors, for example, when null pointer exceptions occur in C++, or the panic function is called in Go or Rust. Previously Envoy did not say anything about the cause of errors, but now you can see the trace which can be used to debug your programs:

Reviewer 2: ~~Now~~

{{< text plain >}}
Function: proxy_on_request_headers failed: Uncaught RuntimeError: unreachable
Proxy-Wasm plugin in-VM backtrace:
  0:  0xdbd - runtime._panic
  1:  0x103ab - main.anotherCalculation
  2:  0x10399 - main.someCalculation
  3:  0xea57 - main.myHeaderHandler
  4:  0xea15 - proxy_on_request_headers
{{< /text >}}

The above is an example output of the stack trace for Go SDK based Wasm extensions. You might notice that you want file names and lines in the trace lines -- that is another huge future work and is currently an open issue since it has something to do with the DWARF format for WebAssembly and the Exception Handling proposal to WebAssembly specification.

### Strace support for Wasm programs

> Also you can see strace equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, we can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example log stream of such strace logs:

Reviewer 2: Make it Also, or just remove the also.

{{< text plain >}}
[host->vm] proxy_on_context_create(2, 1)
[host<-vm] proxy_on_context_create return: void
[host->vm] proxy_on_request_headers(2, 8, 1)
[vm->host] wasi_snapshot_preview1.random_get(86928, 32)
[vm<-host] wasi_snapshot_preview1.random_get return: 0
[vm->host] env.proxy_log(2, 87776, 18)
{{< /text >}}

This is especially useful to debug Wasm program’s execution at runtime, for example, to verify it is not making any malicious system calls.

### Arbitrary Prometheus namespace for in-Wasm metrics

> The next one is about metrics. Wasm extensions have been able to define their own custom metrics and expose them in Envoy just like any other metric, but prior to Istio 1.12, all of these custom metrics are prefixed by `envoy_` Prometheus namespace and users were not be able to have their own namespaces. Now, you can choose whatever namespace you want, and your metrics are exposed in Envoy as-is without being prefixed by `envoy_`.

Reviewer 2: not ~~be~~ able

Note that in order to actually expose these custom metrics, you have to configure [`ProxyConfig.proxyStatsMatcher`](../../../docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-ProxyStatsMatcher) in `meshConfig` for global configuration or in `proxy.istio.io/config` for per proxy configuration. For detail, please refer to [`Envoy Statistics`](../../../docs/ops/configuration/telemetry/envoy-stats/).

## Future work and looking for feedback

> Although we have announced the alpha availability of Wasm plugins, there are still a lot of aspects that are left to be completed. The one important thing is the "Image pull secrets” support in the Wasm API -- this way we would be able to consume the OCI images in the private repository easily in Istio. Others include the first class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.

Reviewer 2: first-class

## New WasmPlugin API

With the new WasmPlugin API in the `extensions.istio.io` namespace, we’re introducing a new high-level API for extending the functionality of the Istio Proxy with custom WebAssembly (Wasm) modules. This effort builds on the excellent work that has gone into the [Proxy-Wasm](https://github.com/proxy-wasm) specification and implementation over the last two years. From now on, you no longer need to use `EnvoyFilter` resources to add custom Wasm modules to your proxies. Instead, you can now use WasmPlugin directly:

{{< text yaml >}}
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: your-filter
spec:
  selector:
    matchLabels:
      app: server
  phase: AUTHN
  priority: 10
  pluginConfig:
    someSetting: true
    someOtherSetting: false
    youNameIt:
    - first
    - second
  url: docker.io/your-org/your-filter:1.0.0
{{< /text >}}

> There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.

Author: Let me add a comment about it

> There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one. The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.

Author: https://github.com/istio/istio.io/pull/10530/commits/b95eef2b7b60f8b707ee274dc2a17ad519de680d done

The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:

> * `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.

Author: yep, IIUC mTLS peer verification will happen before going into filter chains. But then my question is what authentication filters mean here?

> * `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.

Author: @dgn ping

> * `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.

Author: @bianpengyuan can you suggest a change to match the actual behavior?

> * `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.

Reviewer 3: meaning it is executed before any Istio authentication and authorization filters?

> * `AUTHN`: prior to any Istio authentication filters, meaning it is executed even before the remote mTLS certificates are verified.

Author: https://github.com/istio/istio.io/pull/10530/commits/1c1e694f5dae668a1ac8b6fd770cd8be322e5936

`pluginConfig` is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config on `onConfigure` in [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/fd0be8405db25de0264bdb78fae3a82668c03782/proxy_wasm_api.h#L329-L331), `on_configure` in [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or `OnPluginStart` call back in [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).

The `url` field specifies where to pull the Wasm module. You’ll notice that the `url` in this case is a docker URI - this is because apart from loading Wasm modules via HTTP, HTTPS and the local file system (using file://), we are introducing the OCI image format as the preferred mechanism for distributing Wasm modules.

## Wasm image specification

> We believe that containers are the ideal way to store, publish and manage proxy extensions, so we worked with Solo.io to extend their existing Proxy-Wasm container format with a variant that aims to be compatible with all registries and CLI toolchain. Depending on your processes, you can now either build your proxy extension containers using solo’s wasme tool or straight up with your existing container CLI tooling such as Docker CLI or [buildah](https://buildah.io/). Istio supports both variants.

Reviewer 3: I think this needs some clarification. What are the variants Istio supports? I don't think Istio will support Wasme tool, but instead that should be supported by solo.

Author: https://github.com/istio/istio.io/pull/10530/commits/9ec974537a286b58b9822209b40f4f35eff66da7 done

> For detail, please refer to [the link here](https://github.com/solo-io/wasm/tree/master/spec), and learn [how to build OCI images](https://github.com/solo-io/wasm/tree/master/spec#how-can-i-build-images) that are consumable by Istio agent.

Reviewer 3: I think we need to add some guide into istio-ecosystem repo and point there instead. That's the repo where we keep Istio supported guide.

Author: If you are okay, I would be happy to add doc like this https://github.com/mathetake/istio-wasm-api-demo to istio-ecosystem repo, put the link here. sg?

Author: Okay I'm on it. will raise in a hour or so.

Reviewer 3: Thanks! I think it would be nice to just wrap one of the example (either the basic auth one or the plain example extension) about how to build it to an image. It can just be an part of existing guide: https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-a-wasm-extension-with-cpp.md#step-6-push-and-deploy-the-extension

Author: oh I am writing a separate document to put its link here like this:

```
# How to build Wasm OCI images

As of 1.12, Istio has supports for OCI images in its Wasm Plugin mechanism, and users are able to specify
any OCI registries where they store Wasm plugins as containers.

This document describes how to build OCI images which are consumable by Istio.

## Overview
```

so let's make the change to https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/write-a-wasm-extension-with-cpp.md#step-6-push-and-deploy-the-extension after we merge the dedicated document.

Reviewer 3: Yep separated and focusing doc also sounds great to me. Thanks!

Author: https://github.com/istio-ecosystem/wasm-extensions/pull/74 a separated doc for this section

Author: done https://github.com/istio/istio.io/pull/10530/commits/9ec974537a286b58b9822209b40f4f35eff66da7

### Debuggability

Next is the improvement in debuggability. Now the Envoy runtime emits the stack trace of your program when it causes runtime errors, for example, when null pointer exceptions occur in C++, or the panic function is called in Go or Rust. Previously Envoy did not say anything about the cause of errors, but now you can see the trace which can be used to debug your programs:

{{< text plain >}}
Function: proxy_on_request_headers failed: Uncaught RuntimeError: unreachable
Proxy-Wasm plugin in-VM backtrace:
  0:  0xdbd - runtime._panic
  1:  0x103ab - main.anotherCalculation
  2:  0x10399 - main.someCalculation
  3:  0xea57 - main.myHeaderHandler
  4:  0xea15 - proxy_on_request_headers
{{< /text >}}

The above is an example output of the stack trace for Go SDK based Wasm extensions. You might notice that you want file names and lines in the trace lines -- that is another huge future work and is currently an open issue since it has something to do with the DWARF format for WebAssembly and the Exception Handling proposal to WebAssembly specification.

### Strace support for Wasm programs

> Also you can see strace equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, we can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example log stream of such strace logs:

Reviewer 3: Curious: Is this something new added since 1.9? I thought we can already track all ABI calls before.

Author: I think it was between 1.7-1.8, but this section is about the improvements since 1.5 so I think it is fine to explain.

[vm<-host] wasi_snapshot_preview1.random_get return: 0
[vm->host] env.proxy_log(2, 87776, 18)
{{< /text >}}

This is especially useful to debug Wasm program’s execution at runtime, for example, to verify it is not making any malicious system calls.

### Arbitrary Prometheus namespace for in-Wasm metrics

The next one is about metrics. Wasm extensions have been able to define their own custom metrics and expose them in Envoy just like any other metric, but prior to Istio 1.12, all of these custom metrics are prefixed by `envoy_` Prometheus namespace and users were not be able to have their own namespaces. Now, you can choose whatever namespace you want, and your metrics are exposed in Envoy as-is without being prefixed by `envoy_`.

Note that in order to actually expose these custom metrics, you have to configure [`ProxyConfig.proxyStatsMatcher`](../../../docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-ProxyStatsMatcher) in `meshConfig` for global configuration or in `proxy.istio.io/config` for per proxy configuration. For detail, please refer to [`Envoy Statistics`](../../../docs/ops/configuration/telemetry/envoy-stats/).

### Bug fixes, and more

> In addition to all of the above, we have fixed tons of bugs found in Envoy and refactored the original code. Notably now all of the Istio specific Wasm related codes in proxy have been removed, meaning that Istio telemetry and any other Proxy-Wasm based Istio extensions just depend on Proxy-Wasm ABI and the upstream Envoy implementation. This proves that the direction of the Proxy-Wasm project towards defining generic Wasm ABI to extend network proxies is on the right track.

Author: deleted this subsection!

---
title: Announcing the alpha availability of WebAssembly Plugins

> description: Introduction of Wasm Plugin API, and updates on Envoy and Istio's Wasm based plugins.

Reviewer 4: ```suggestion
description: Introduction to the new Wasm Plugin API and updates to the Wasm-based plugin support in Envoy and Istio.
```

> publishdate: 2021-11-25

Reviewer 4: ```suggestion
publishdate: 2021-12-16
```

attribution: "Daniel Grimm (Red Hat), Pengyuan Bian (Google), Takeshi Yoneda (Tetrate)"

keywords: [wasm,extensibility,WebAssembly]
---

> [Istio 1.9 introduced](../wasm-progress/) experimental support for WebAssembly (Wasm) module distribution and a Wasm extensions ecosystem repository with canonical examples and use cases for extension development. Over the past 9 months, the Istio, Envoy, and Proxy-Wasm communities have continued our joint efforts to make Wasm extensibility stable, reliable, and easy to adopt, and we are pleased to announce Alpha support for Wasm extensibility at Istio 1.12 release! Let’s walk through the updates to Wasm support through the Istio 1.12 release.

Reviewer 4: ```suggestion
[Istio 1.9 introduced](../wasm-progress/) experimental support for WebAssembly (Wasm) module distribution and a Wasm extensions ecosystem repository with canonical examples and use cases for extension development. Over the past 9 months, the Istio, Envoy, and Proxy-Wasm communities have continued our joint effort to make Wasm extensibility stable, reliable, and easy to adopt, and we are pleased to announce Alpha support for Wasm extensibility in Istio 1.12! In the following sections, we'll walk through the updates that have been made to the Wasm support for the 1.12 release.
```

> [At the Istio 1.9 release](../wasm-progress/), we introduced experimental support for WebAssembly(Wasm) module distribution and Wasm extensions ecosystem repository for canonical examples and use cases of Wasm extension development. Over the past 9 months, the Istio, Envoy, and Proxy-Wasm communities have continued our joint efforts to make Wasm extensibility stable, reliable, and easy to adopt, and we are pleased to announce Alpha support for Wasm extensibility at Istio 1.12 release! Let’s walk through the updates to Wasm support through the Istio 1.12 release.

Reviewer 4: ```suggestion
[Istio 1.9 introduced](../wasm-progress/) experimental support for WebAssembly (Wasm) module distribution and a Wasm extensions ecosystem repository with canonical examples and use cases for extension development. Over the past 9 months, the Istio, Envoy, and Proxy-Wasm communities have continued our joint efforts to make Wasm extensibility stable, reliable, and easy to adopt, and we are pleased to announce Alpha support for Wasm extensibility at Istio 1.12 release! Let’s walk through the updates to Wasm support through the Istio 1.12 release.
```

## New WasmPlugin API

> With the new WasmPlugin API in the `extensions.istio.io` namespace, we’re introducing a new high-level API for extending the functionality of the Istio Proxy with custom Wasm modules. This effort builds on the excellent work that has gone into the [Proxy-Wasm](https://github.com/proxy-wasm) specification and implementation over the last two years. From now on, you no longer need to use `EnvoyFilter` resources to add custom Wasm modules to your proxies. Instead, you can now use WasmPlugin directly:

Reviewer 4: ```suggestion
With the new `WasmPlugin` CRD in the `extensions.istio.io` namespace, we’re introducing a new high-level API for extending the functionality of the Istio proxy with custom Wasm modules. This effort builds on the excellent work that has gone into the [Proxy-Wasm](https://github.com/proxy-wasm) specification and implementation over the last two years. From now on, you no longer need to use `EnvoyFilter` resources to add custom Wasm modules to your proxies. Instead, you can now use a `WasmPlugin` resource:
```

{{< text yaml >}}
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: your-filter
spec:
  selector:
    matchLabels:
      app: server
  phase: AUTHN
  priority: 10
  pluginConfig:
    someSetting: true
    someOtherSetting: false
    youNameIt:
    - first
    - second
  url: docker.io/your-org/your-filter:1.0.0
{{< /text >}}

> There are a lot of similarities and a few differences between WasmPlugin and `EnvoyFilter`, so let’s go through the fields one by one.

Reviewer 4: ```suggestion
There are a lot of similarities and a few differences between `WasmPlugin` and `EnvoyFilter`, so let’s go through the fields one by one.
```

The above example deploys a Wasm module to all workloads (including gateway pods) that match it’s `selector` - this very much works the same as in an `EnvoyFilter`.

The next field below that is the `phase`. This determines where in the proxy’s filter chain the Wasm module will be injected. We have defined four distinct phases for injection:

* `AUTHN`: prior to any Istio authentication and authorization filters.

> * `AUTHZ`: after the Istio authentication filters and before any first-class authorization filters, i.e. before AuthorizationPolicies have been applied.

Reviewer 4: ```suggestion
* `AUTHZ`: after the Istio authentication filters and before any first-class authorization filters, i.e., before AuthorizationPolicies have been applied.
```

* `STATS`: after all authorization filters and prior to the Istio stats filter.

> * `UNSPECIFIED_PHASE`: let the control plane decide where to insert. This will generally be at the end of the filter chain, right before the Router. This is the default value for this `phase` field.

Reviewer 4: ```suggestion
* `UNSPECIFIED_PHASE`: let the control plane decide where to insert. This will generally be at the end of the filter chain, right before the router. This is the default value for this `phase` field.
```

> `pluginConfig` is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config on `onConfigure` in [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/fd0be8405db25de0264bdb78fae3a82668c03782/proxy_wasm_api.h#L329-L331), `on_configure` in [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or `OnPluginStart` call back in [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).

Reviewer 4: ```suggestion
The `pluginConfig` field is used for configuring your Wasm plugin. Whatever you put into this field will be encoded in JSON and passed on to your filter, where you can access it in the configuration callback of the Proxy-Wasm SDKs. For example, you can retrieve the config with `onConfigure` in the [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/fd0be8405db25de0264bdb78fae3a82668c03782/proxy_wasm_api.h#L329-L331), `on_configure` in the [Rust SDK](https://github.com/proxy-wasm/proxy-wasm-rust-sdk/blob/v0.1.4/src/dispatcher.rs#L255) or the `OnPluginStart` call back in the [Go SDK](https://github.com/tetratelabs/proxy-wasm-go-sdk/blob/v0.15.0/proxywasm/types/context.go#L74).
```

The `url` field specifies where to pull the Wasm module. You’ll notice that the `url` in this case is a docker URI - this is because apart from loading Wasm modules via HTTP, HTTPS and the local file system (using file://), we are introducing the OCI image format as the preferred mechanism for distributing Wasm modules.

One last thing to note is currently the Wasm Plugin API only applies to inbound HTTP filter chains.

> Adding support for network filters and outbound traffics is the future work.

Reviewer 4: ```suggestion
Support for network filters and outbound traffic will be added in the future.
```

## Wasm image specification

We believe that containers are the ideal way to store, publish and manage proxy extensions, so we worked with Solo.io to extend their existing Proxy-Wasm container format with a variant that aims to be compatible with all registries and the CLI toolchain. Depending on your processes, you can now build your proxy extension containers using your existing container CLI tooling such as Docker CLI or [buildah](https://buildah.io/).

> To learn how to build OCI images, please refer to [this link](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md).

Reviewer 4: ```suggestion
To learn how to build OCI images, please refer to [these instructions](https://github.com/istio-ecosystem/wasm-extensions/blob/master/doc/how-to-build-oci-images.md).
```

## Image fetcher in Istio agent

> As of 1.9, Istio-agent provides a reliable solution for loading Wasm binaries fetched from remote HTTP sources configured in the `EnvoyFilters` by leveraging the xDS proxy inside istio-agent and Envoy’s Extension Configuration Discovery Service (ECDS). The same mechanism applies for the new Wasm API implementation in Istio 1.12, and you can use HTTP remote resources reliably without the concern that Envoy might get stuck in bad configurations when the remote fetch fails.

Reviewer 4: ```suggestion
Since Istio 1.9, Istio-agent has provided a reliable solution for loading Wasm binaries, fetched from remote HTTP sources configured in the `EnvoyFilters`, by leveraging the xDS proxy inside istio-agent and Envoy’s Extension Configuration Discovery Service (ECDS). The same mechanism applies for the new Wasm API implementation in Istio 1.12. You can use HTTP remote resources reliably without concern that Envoy might get stuck with a bad configuration when a remote fetch fails.
```

> In addition, Istio 1.12 expands this capability to the Wasm OCI images. That means the Istio agent is now able to fetch Wasm images from any OCI registry including Docker Hub, Google Container Registry(GCR), Amazon Elastic Container Registry (Amazon ECR), etc. After fetching images, Istio-agent extracts and caches Wasm binaries from them, and then inserts them into Envoy filter chains.

Reviewer 4: ```suggestion
In addition, Istio 1.12 expands this capability to Wasm OCI images. This means the Istio-agent is now able to fetch Wasm images from any OCI registry including Docker Hub, Google Container Registry(GCR), Amazon Elastic Container Registry (Amazon ECR), etc. After fetching images, Istio-agent extracts and caches Wasm binaries from them, and then inserts them into the Envoy filter chains.
```

{{< image width="75%"
    link="./istio-agent-architecture.svg"
    alt="Remote Wasm module fetch flow"
    caption="Remote Wasm module fetch flow"
    >}}

## Improvements in Envoy Wasm runtime

> The Wasm runtime powered by V8 in Envoy has been shipped as of Istio 1.5 and there has been a lot of improvements since then.

Reviewer 4: ```suggestion
The Wasm runtime powered by V8 in Envoy has been shipped since Istio 1.5 and there have been a lot of improvements since then.
```

### WASI supports

> First, some of the WASI (WebAssembly System Interface) system calls are now supported. For example, `clock_time_get` system call can be made from Wasm programs, which means that you can use `std::time::SystemTime::now()` in Rust or `time.Now().UnixNano()` in Go for Envoy Wasm extensions just like any other native platform. Another example is that `random_get` is supported by Envoy now, so the "crypto/rand" package is available in Go as a cryptographically secure random number generator. We are currently actively looking into file system support as we have seen the requests for reading and writing local files from Wasm programs running in Envoy.

Reviewer 4: ```suggestion
First, some of the WASI (WebAssembly System Interface) system calls are now supported. For example, the `clock_time_get` system call can be made from Wasm programs so you can use `std::time::SystemTime::now()` in Rust or `time.Now().UnixNano()` in Go in your Envoy Wasm extensions, just like any other native platform. Another example is `random_get` is now supported by Envoy, so the "crypto/rand" package is available in Go as a cryptographically secure random number generator. We are also currently looking into file system support as we have seen requests for reading and writing local files from Wasm programs running in Envoy.
```

### Debuggability

> Next is the improvement in debuggability. Now the Envoy runtime emits the stack trace of your program when it causes runtime errors, for example, when null pointer exceptions occur in C++, or the panic function is called in Go or Rust. Previously Envoy did not say anything about the cause of errors, but you can see the trace which can be used to debug your programs:

Reviewer 4: ```suggestion
Next is the improvement in debuggability. The Envoy runtime now emits the stack trace of your program when it causes runtime errors, for example, when null pointer exceptions occur in C++ or the panic function is called in Go or Rust. While Envoy error messages did not previously include anything about the cause, they now show the trace which you can use to debug your program:
```

{{< text plain >}}
Function: proxy_on_request_headers failed: Uncaught RuntimeError: unreachable
Proxy-Wasm plugin in-VM backtrace:
  0:  0xdbd - runtime._panic
  1:  0x103ab - main.anotherCalculation
  2:  0x10399 - main.someCalculation
  3:  0xea57 - main.myHeaderHandler
  4:  0xea15 - proxy_on_request_headers
{{< /text >}}

> The above is an example output of the stack trace for Go SDK based Wasm extensions. You might notice that you want file names and lines in the trace lines -- that is another huge future work and is currently an open issue since it has something to do with the DWARF format for WebAssembly and the Exception Handling proposal to WebAssembly specification.

Reviewer 4: ```suggestion
The above is an example stack trace from a Go-SDK based Wasm extension. You might notice that the output does not include file names and line numbers in the trace. This is an important future work item and open issue related to the DWARF format for WebAssembly and the Exception Handling proposal for the WebAssembly specification.
```

### Strace support for Wasm programs

> You can see strace equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, we can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example log stream of such strace logs:

Reviewer 4: ```suggestion
You can see `strace` equivalent logs emitted by Envoy. With Istio proxy’s component log level `wasm:trace`, you can observe all the system calls and Proxy-Wasm ABI calls that go across the boundary between Wasm virtual machines and Envoy. The following is an example of such an `strace` log stream:
```

{{< text plain >}}
[host->vm] proxy_on_context_create(2, 1)
[host<-vm] proxy_on_context_create return: void
[host->vm] proxy_on_request_headers(2, 8, 1)
[vm->host] wasi_snapshot_preview1.random_get(86928, 32)
[vm<-host] wasi_snapshot_preview1.random_get return: 0
[vm->host] env.proxy_log(2, 87776, 18)
{{< /text >}}

> This is especially useful to debug Wasm program’s execution at runtime, for example, to verify it is not making any malicious system calls.

Reviewer 4: ```suggestion
This is especially useful to debug a Wasm program's execution at runtime, for example, to verify it is not making any malicious system calls.
```

### Arbitrary Prometheus namespace for in-Wasm metrics

> The next one is about metrics. Wasm extensions have been able to define their own custom metrics and expose them in Envoy just like any other metric, but prior to Istio 1.12, all of these custom metrics are prefixed by `envoy_` Prometheus namespace and users were not able to have their own namespaces. Now, you can choose whatever namespace you want, and your metrics are exposed in Envoy as-is without being prefixed by `envoy_`.

Reviewer 4: ```suggestion
The last update is about metrics. Wasm extensions have been able to define their own custom metrics and expose them in Envoy, just like any other metric, but prior to Istio 1.12, all of these custom metrics were prefixed with the `envoy_` Prometheus namespace and users were not able to use their own namespaces. Now, you can choose whatever namespace you want, and your metrics will be exposed in Envoy as-is, without being prefixed by `envoy_`.
```

Note that in order to actually expose these custom metrics, you have to configure [`ProxyConfig.proxyStatsMatcher`](../../../docs/reference/config/istio.mesh.v1alpha1/#ProxyConfig-ProxyStatsMatcher) in `meshConfig` for global configuration or in `proxy.istio.io/config` for per proxy configuration. For detail, please refer to [`Envoy Statistics`](../../../docs/ops/configuration/telemetry/envoy-stats/).

## Future work and looking for feedback

> Although we have announced the alpha availability of Wasm plugins, there are still a lot of aspects that are left to be completed. The one important thing is the "Image pull secrets” support in the Wasm API -- this way we would be able to consume the OCI images in the private repository easily in Istio. Others include the first-class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.

Reviewer 4: ```suggestion
Although we have announced the alpha availability of Wasm plugins, there are still a lot of work left to be done. One important work item is "Image pull secrets” support in the Wasm API which will allow you to easily consume OCI images in a private repository. Others include first-class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.
```

> Although we have announced the alpha availability of Wasm plugins, there are still a lot of work left to be done. One important work item is "Image pull secrets” support in the Wasm API which will allow you to easily consume OCI images in a private repository. Others include first-class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.

Reviewer 4: ```suggestion
Although we have announced the alpha availability of Wasm plugins, there are still a lot of work left to be done. One important work item is "Image pull secrets” support in the Wasm API which will allow you to easily consume OCI images in a private repository. Others include first-class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.
```
```suggestion
Although we have announced the alpha availability of Wasm plugins, there is still a lot of work left to be done. One important work item is "Image pull secrets” support in the Wasm API which will allow you to easily consume OCI images in a private repository. Others include first-class support for L4 filters, signature verification of Wasm binaries, runtime improvements in Envoy, Proxy-Wasm SDK improvements, documentation, etc.
```

> That means, this is just the beginning of the 1st-class Wasm support in Istio. We would love to hear feedback from users so that we could improve the developer experience with the Wasm plugins!

Reviewer 4: ```suggestion
This is just the beginning of our plan to provide 1st-class Wasm support in Istio. We would love to hear your feedback so that we can improve the developer experience using Wasm plugins, in future releases of Istio!
```
